Feature/ms deal/created deal microservice

futind opened the pull request:
Сделал микросервис, написал документацию, сделал логирование.
Написал несколько тестов (покрытие 63%).
Поправил одно недопонимание моё в калькуляторе.

Креды захардкожены пока что.

futind committed:
Initialized the Calculator microservice.

futind committed:
Created the MS-Calculator microservice.
Defined the DTOs, defined and implemented endpoints for "/offers" and "/calc".

futind committed:
Fixes, Logging, Documentation and Unit-tests.
Fixed everything that was pointed out in the previous PR. Added logging, documentation using Swagger and Unit-tests using JUnit.

futind committed:
Annotation changes and Swagger examples

futind committed:
Merge pull request #2 from futind/feature/MS-Calculator/CreatedTheMicroservice

Created the MS-Calculator microservice.

futind committed:
Created the Deal microservice/Fixes in Calculator
Created the microservice, including logging, API documentation using swagger, some (a few) tests.
Also fixed some things in MS-calculator which were needed for Deal to work as intended.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/dtos/ScoringDataDto.java:
```diff
@@ -0,0 +1,107 @@
+    @NotNull
+    @Size(min = 2, max = 30)
+    @Pattern(regexp = "^[a-zA-Z]+$")
+    @Schema(description = "Имя клиента", example = "Джон")
```
Прошу прощения, что только сейчас заметил, но тут медвежья услуга: ты вписал Джон, чтобы Сваггер это дело подставил. НО! У тебя валидация только английские символы пропускает :)

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/dtos/ScoringDataDto.java:
```diff
@@ -0,0 +1,107 @@
+    @NotNull
+    @Valid
+    @Schema(description = "Данные о занятости клиента", example = "SELF_EMPLOYED")
+    private EmploymentDto employment;
```
Тут описка. EmploymentDto не может быть example = "SELF_EMPLOYED" :) , т.к. это структура, а не поле

KonstantinSmelov commented on MS-Deal/src/main/java/ru/neoflex/msdeal/controller/DealController.java:
```diff
@@ -0,0 +1,205 @@
+                    по мере уменьшения процентной ставки.".
+                    """
+    )
+    @PostMapping("/statement")
```
Прошлый раз не стал делать предложение, чтобы не тормозить переход к следующему заданию. Но сейчас хотел бы всё-таки предложить.
Описания для Сваггера крайне затрудняют чтение контроллера.
Можно, к примеру, сделать интерфейс для класса-контроллера, который будет содержать методы контроллера. В этом интерфейсе ты это всё для Сваггера опишешь (как сейчас в классе-контроллере), а в классе-контроллере просто будешь имплементировать этот интерфейс. 
 Таким образом контроллере будет только бизнес-логика, без огромных полотен для Сваггера (которые будут описаны в интерфейсе).
 
Это не единственный такой момент.

KonstantinSmelov commented on MS-Deal/src/main/java/ru/neoflex/msdeal/controller/DealController.java:
```diff
@@ -0,0 +1,205 @@
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(request)
+                .retrieve()
+                .body(new ParameterizedTypeReference<List<LoanOfferDto>>() {});
```
Слишком много логики для контроллера. 
Тут, в данном случае, должен вызываться один метод, который внутри себя делает всё что надо.

Но тупо переносить всё в сервисный метод тоже не надо. Вызовы restClient надо в отдельную функцию вынести, чтобы вызывать что-то типа getScoring() внутри этого метода. И т.п.

И касательно сервисных методов, надо прочувствовать балланс. 
С одной стороны чтобы не было одного сервисного метода, который делает всё. Но с другой стороны не надо делать миллион сервисных методов на каждый чих.
Это касается и других подобных мест в коде.

P.S. Почему действие, которое может вызвать эксепшн тут не ловится? А ниже, в методе finishRegistration(), запрос к restClient уже в блоке try? Почему такое непостоянство? Или это ты визуализируешь рост своих скиллов в реальном времени? :)

KonstantinSmelov commented on MS-Deal/src/main/java/ru/neoflex/msdeal/controller/DealController.java:
```diff
@@ -0,0 +1,205 @@
+                .retrieve()
+                .body(new ParameterizedTypeReference<List<LoanOfferDto>>() {});
+
+        assert offers != null;
```
У меня IDEA пишет, что бессмысленный ассёрт, т.к. offers всегда не null.
Очевидно, она имеет в виду, что если метод выше (restClient) отработал без ошибки, то offers будет создан. Возможно путой, но создан (т.е. не null по любому). А если restClient отвалится во время запроса, то до этой строчки код и не дойдёт.

KonstantinSmelov commented on MS-Deal/src/main/java/ru/neoflex/msdeal/controller/DealController.java:
```diff
@@ -0,0 +1,205 @@
+
+        for (LoanOfferDto offer : offers) {
+            offer.setStatementId(statementEntity.getStatementId());
+        }
```
Ну сам же бог велел использовать тут стрим для прохождения по всем элементам. Ты же ими пользуешься в коде, я видел. Вот тут прямо надо-надо.

KonstantinSmelov commented on MS-Deal/src/main/java/ru/neoflex/msdeal/service/ClientService.java:
```diff
@@ -0,0 +1,70 @@
+import ru.neoflex.msdeal.repository.ClientRepository;
+
+import java.util.UUID;
+
```
Тебе надо почитать и понять, когда и где надо использовать аннотацию @ Transactional.

KonstantinSmelov commented on MS-Deal/src/main/resources/application.properties:
```diff
@@ -0,0 +1,15 @@
+spring.datasource.driver-class-name=org.postgresql.Driver
+spring.datasource.url=jdbc:postgresql://localhost:5432/loan_issuer_database
+spring.datasource.username=postgres
+spring.datasource.password=admin
```
Касательно сохранения секретов в гит.
Простых вариантов несколько.

Например, можно использовать два профиля. Т.е. основной профиль вот этот вот. Где ты обозначаешь, к примеру, spring.datasource.password = , но не присваиваешь значения.
И ещё один профиль, который Спринг будет применять после и поверх основного и дополнять его. В частности, проставлять пароль для spring.datasource.password = admin. 
Этот второй профиль с живым паролем, ты добавляешь в гитигнор, и храниться он у тебя только локально. А в основном профиле указываешь, чтобы подтягивался ещё один профиль.

KonstantinSmelov commented on MS-Deal/src/test/java/ru/neoflex/msdeal/service/ClientServiceTest.java:
```diff
@@ -0,0 +1,111 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+@ActiveProfiles("test")
+@SpringBootTest
```
В тестах сервиса, ты тестируешь логику сервиса, а не взаимодействие этого сервиса с БД (в данном случае). Тебе контекст спринга тут поднимать не надо.

В классе этого метода надо использовать аннотации InjectMocks и Mock. Почитай про них.

Логика на примере метода createClientWithRequest():

Этот метод принимает объект (LoanStatementRequestDto), на основании этого объекта строит другой объект (ClientEntity), сохраняет этот ClientEntity в БД и заодно возвращает его.

Посему, ты должен в тесте проверить, что до метода save() дошёл объект такой, какой должен был дойти, что все поля правильно проинициализировались в нём и вернулось то, что должно вернуться.

Т.к. в БД мы писать ничего не собираемся в рамках этих юнит-тестов, то метод save(), который вызывается внутри этого метода, мы будем мокать. И опишем его ожидаемое поведение с помощью when(clientRepository.save(...)).thenReturn(..,.,);
Потом выполним тестируемый метод createClientWithRequest(...) с сверим, что он вернул то, что должен был вернуть при определённых входных параметрах.

Сейчас, я думаю, тебя смущает максимальная простота логики в твоих методах. Поэтому твой мозг отторгает подобные тесты. Т.к. если, грубо говоря, метод принимает цифру 1, а потом возвращает эту цифру, то тут нечего тестировать. Слишком прозрачно.
В реальных проектах методы со значительно сложной логикой будут, а подход к их тестированию ты уже будешь понимать, т.к. набил руку на этом проекте.

futind committed:
Fixes
Added more tests, fixed the mistakes that had been pointed out earlier.
Also added more meaningful tests.

futind committed:
Fixed Swagger documentation to show only ms_deal
